### PR TITLE
allow passing a passphrase for SSL certs (from command line or file)

### PR DESCRIPTION
**Pull request by moneromooo-monero**

### CODE REVIEW

**Comment by omartijn** [comment]
I wonder whether we should have the option of a passphrase file. It seems to sort of negate the benefits of having a passphrase on your certificates at all.

**Comment by moneromooo-monero** [comment]
That's a fair point. It's what we do for monero-wallet-rpc though. I don't really have a better idea for unattended runs.

**Comment by omartijn** [comment]
Well, I see two alternative options. One would be to add it as an cli option, but that has very obvious drawbacks. Anybody with access to the system could use `ps` to see your passphrase.

Another option is to let users pipe in the passphrase on stdin. This wouldn't have to be visible in the process list, depending on where they get it from. In case they want to keep using a passphrase file, it could be as easy as `cat passphrase | monero-wallet-rpc ...`. It also allows people to write scripts retrieving the passphrase from other services (like Vault, or maybe a database).

However, since you mention we do it this way in the wallet-rpc already, for consistency, I'd accept that we do it this way for now. I think it'd be a good candidate to change later on.

File: contrib/epee/src/net_ssl.cpp [hunk 1]
```diff
@@ -254,6 +254,10 @@ boost::asio::ssl::context ssl_options_t::create_context() const
 
 void ssl_authentication_t::use_ssl_certificate(boost::asio::ssl::context &ssl_context) const
 {
+  if (!private_key_passphrase.empty())
+    ssl_context.set_password_callback([this](size_t max_length, boost::asio::ssl::context::password_purpose purpose)->std::string{
```
Comment (vtnerd) [inline contrib/epee/src/net_ssl.cpp]: Is the `wipeable_string` instance really that useful if a copy is being made anyway?

File: src/rpc/core_rpc_server.cpp [hunk 2]
```diff
@@ -91,6 +92,8 @@ namespace cryptonote
     command_line::add_arg(desc, arg_restricted_rpc);
     command_line::add_arg(desc, arg_rpc_ssl);
     command_line::add_arg(desc, arg_rpc_ssl_private_key);
+    command_line::add_arg(desc, arg_rpc_ssl_private_key_passphrase);
+    command_line::add_arg(desc, arg_rpc_ssl_private_key_passphrase_file);
```
Comment (vtnerd) [inline src/rpc/core_rpc_server.cpp]: As per @omartijn comments, I don't see a reason to support/maintain a passphrase file specifically in the core rpc code. I don't see a good reason for the command line-switch passphrase either, although people aggressively running in pid namespaces (i.e. containers) can hide the argument list in some cases. It seems like a `stdin` or prompt would be the only useful scenario.

File: src/rpc/core_rpc_server.cpp [hunk 3]
```diff
@@ -164,8 +167,27 @@ namespace cryptonote
...
+      std::string passphrase;
+      if (!epee::file_io_utils::load_file_to_string(passphrase_file, passphrase))
+      {
+        MERROR("Failed to load passphrase");
+        return false;
+      }
+
+      // Remove line breaks the user might have inserted
+      boost::trim_right_if(passphrase, boost::is_any_of("\r\n"));
+      private_key_passphrase = passphrase;
```
Comment (vtnerd) [inline src/rpc/core_rpc_server.cpp]: A move constructor exists, `private_key_passphrase = std::move(passphrase);`. The `memwipe` can still be done, but the string should be empty by then.

File: src/wallet/wallet2.cpp [hunk 4]
```diff
@@ -355,8 +359,24 @@ std::unique_ptr<tools::wallet2> make_basic(const boost::program_options::variabl
        tools::wallet2::tr("Invalid argument for ") + std::string(opts.daemon_ssl.name));
   }
 
+  epee::wipeable_string private_key_passphrase;
+  std::string passphrase_file = command_line::get_arg(vm, opts.daemon_ssl_private_key_passphrase_file);
```
Comment (vtnerd) [inline src/wallet/wallet2.cpp]: Again, I don't see a "value-add" in this situation.

File: src/wallet/wallet_rpc_server.cpp [hunk 5]
```diff
@@ -270,8 +274,27 @@ namespace tools
       }
     }
 
+    epee::wipeable_string private_key_passphrase;
+    std::string passphrase_file = command_line::get_arg(vm, arg_rpc_ssl_private_key_passphrase_file);
+    if (!passphrase_file.empty())
```
Comment (vtnerd) [inline src/wallet/wallet_rpc_server.cpp]: This is the useful place ... kind of (consistency with other arguments) ?

**Comment by moneromooo-monero** [comment]
Well, WTF.

**Pull request closed without merging**